src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java

```diff
@@ -0,0 +1,328 @@
+package org.broadinstitute.hellbender.tools.spark.bwa;
+
+import com.github.lindenb.jbwa.jni.BwaIndex;
+import com.github.lindenb.jbwa.jni.BwaMem;
+import com.github.lindenb.jbwa.jni.ShortRead;
+import com.google.common.base.Stopwatch;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.UnmodifiableIterator;
+import htsjdk.samtools.*;
+import htsjdk.samtools.reference.ReferenceSequence;
+import htsjdk.samtools.reference.ReferenceSequenceFile;
+import htsjdk.samtools.reference.ReferenceSequenceFileFactory;
+import htsjdk.samtools.util.StringUtil;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.broadinstitute.hellbender.cmdline.Argument;
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.cmdline.programgroups.SparkProgramGroup;
+import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
+import org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSink;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadsWriteFormat;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.seqdoop.hadoop_bam.BAMInputFormat;
+import scala.Tuple2;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.*;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.Function;
+
+import static org.broadinstitute.hellbender.utils.Utils.calcMD5;
+
+@CommandLineProgramProperties(summary = "Runs BWA",
+        oneLineSummary = "BWA on Spark",
+        programGroup = SparkProgramGroup.class)
+public final class BwaSpark extends GATKSparkTool {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger log = LogManager.getLogger(BwaSpark.class);
+
+    private static ThreadLocal<Stopwatch> stopwatchThreadLocal = new ThreadLocal<Stopwatch>() {
+        @Override
+        protected Stopwatch initialValue() {
+            return Stopwatch.createUnstarted();
+        }
+    };
+
+    private static Set<Stopwatch> stopwatches = new HashSet<>();
+
+    @Argument(doc = "the reference", shortName = "ref",
+            fullName = "ref", optional = false)
+    private String ref;
```

Comment:
use `StandardArgumentDefinitions.REFERENCE_LONG_NAME` and StandardArgumentDefinitions.REFERENCE_SHORT_NAME` for the reference argument